### DISCUSSION TITLE
NDVI calculated from sentinel-2-l2a is very high

### DISCUSSION THREAD

**Discussion by mehran66** [microsoft/PlanetaryComputer#285] [Q&A]
I have been using sentinel-2-l2a to calcuate NDVI and noticed that the values are quite high after comparing to the corresponding NDVIs calcuated from Niap. Is there any considerations for the NDVIs from sentinel-2-l2a. Here is a sample code:

        bbox = (-117.23897137214534,
         33.10812294988965,
         -117.23146862654204,
         33.11443704850675)
        
        stac = pystac_client.Client.open(
            "https://planetarycomputer.microsoft.com/api/stac/v1",
            modifier=planetary_computer.sign_inplace,
        )
        
        search = stac.search(
            bbox=bbox,
            datetime="2016-06-28/2017-08-28",
            collections=["sentinel-2-l2a"],
            query={"eo:cloud_cover": {"lt": 25}},
        )
        
        items = search.item_collection()
        
        
        data = (
            stackstac.stack(
                items,
                assets=["B02", "B03", "B04", "B08"], 
                chunksize=2304,
                resolution=10,
                bounds_latlon=bbox,
            )
            .where(lambda x: x > 0, other=np.nan)  # sentinel-2 uses 0 as nodata
            .assign_coords(band=lambda x: x.common_name.rename("band"))  # use common names
        )
        
        data = data.persist()
        
        red, nir = data.sel(band="red"), data.sel(band="nir")
        ndvi = ((nir - red) / (nir + red)).astype('float32')
        median_ndvi = ndvi.median(dim="time").compute()

**Comment by TomAugspurger** [comment]
I'm not sure if it's meaningful to compare NDVI from two different sensors. If one of the sensors captures one of the bands in a different wavelength, it would lead to different values for NDVI. Our [sentinel-2 docs](https://planetarycomputer.microsoft.com/dataset/sentinel-2-l2a) has the wavelengths. We don't for NAIP, but maybe their documentation has that.

**Comment by mehran66** [comment]
Thanks. It's correct that that direct comparisons of NDVIs calculated from diferent sensors is not meaningful, but I found it suspicious that the median and mean NDVI of sentinel is larger than Naip by 0.2-0.4. looking at the NDVI rasters, the results of naip is more realistic and I guessed that sth wrong has happend to the sentinel bands.

**Reply by Kongstad** [reply]
The code you provided shows dates from 2016-2017, so this may not be relevant to you. But just incase the above is just an example and you are working with data across 2021/2022, then this may be helpful to you.
https://forum.step.esa.int/t/changes-in-band-data-after-25-jan-2022-baseline-04-00-harmonizevalues-sentinel-2-l2a-snappy/36270

**Reply by mehran66** [reply]
Thanks for sharing this @Kongstad 
Based on my experience, the NDVI calculated from sentinel-2-l2a is always higher that the ones calculated from Naip (about 0.2). I am not sure if that much difference is normal as it seems more like a systematic bias.

**Reply by KBodolai** [reply]
How does it look if you compare it to the NDVI calculated from Landsat? it could give a decent idea of which is the outlier here?

**Reply by mehran66** [reply]
@KBodolai running that for a random location ... the median of NDVI for an area is
sentinel: 0.32
Landsat: 0.13
Naip: 0.23

I thought that maybe it is not fair to compare to Naip given that it is captured once every few years and probably green seasons. But now I see that it is even worst comparing to Landsat.

**Reply by KBodolai** [reply]
@mehran66 , this is indeed odd. Difficutl to tell what's wrong without a specific example, but just checking, have you applied the scale factors to convert the Landsat data back to reflectance (as described [here](https://www.usgs.gov/faqs/how-do-i-use-a-scale-factor-landsat-level-2-science-products))? 

The offset factor introduces a bias that 'squishes' the NDVI range, since the scaling is of the form:

$$
\text{digital number} = \frac{\text{reflectance} - \text{offset}}{\text{scale factor}} 
$$

for each band, so NDVI in digital number would be biased with respect to the reflectance values:

$$
NDVI_{dn} = \frac{N - R}{N+R-2*\text{offset}}
$$

which is always of smaller magnitude than 

$$
NDVI_{\text{refl}} = \frac{N-R}{N+R}
$$

since the offset is -0.2. (N and R are the values for the near infrared and red bands in reflectance)
